src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -1129,44 +1142,15 @@ public string DnsSafeHost
 
                 EnsureHostString(false);
 
-                if (_info.DnsSafeHost != null)
-                {
-                    // Cached
-                    return _info.DnsSafeHost;
-                }
-                else if (_info.Host!.Length == 0)
-                {
-                    // Empty host, no possible processing
-                    return string.Empty;
-                }
-
-                // Special case, will include ScopeID and strip [] around IPv6
-                // This will also unescape the host string
-                string ret = _info.Host;
-
-                if (HostType == Flags.IPv6HostType)
+                Flags hostType = HostType;
+                if (hostType == Flags.IPv6HostType || (hostType == Flags.BasicHostType && InFact(Flags.HostNotCanonical | Flags.E_HostNotCanonical)))
```

Comment:
Do we have enough test coverage in place to fully validate the change?

Reply:
I believe so, but added some more tests